onnx/defs/math/defs.cc

```diff
@@ -2006,6 +2007,301 @@ ONNX_OPERATOR_SET_SCHEMA(
+        Byte 1: [ . . . . . . . C ] -> C[0]
+      ```
+  The last `uint_8` byte may have some bits unused.
+
+Input `scales` is stored in same type as original type of B(`float32`, `float16`) with shape like:
+`[N * n_blocks_per_col]`
+
+Input `zero_points` is stored as `uint8_t` or the same type as `A`. It has the same packing method as input `B`.
+  - If `zero_points` type is `uint8_t`, its shape is `[N * CeilDiv(n_blocks_per_col * bits, 8)]`.
+  - If `zero_points` has same type as `A`, it's not packed and has the same shape as `scales`, `[N * n_blocks_per_col]`.
+  - If `zero_points` is not provided then zero_points will be set to `2^(bits - 1)`.
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    MatMulNBits,
+    23,
+    OpSchema()
+        .SetDoc(MatMulNBits_ver23_doc)
+        .Attr(
+            "K",
+            "Size of each input feature. If not present, should be inferred from input A.",
+            AttributeProto::INT,
+            OPTIONAL_VALUE)
+        .Attr(
+            "N",
+            "Size of each output feature. If not present, should be inferred from input B.",
+            AttributeProto::INT,
+            OPTIONAL_VALUE)
+        .Attr(
+            "accuracy_level",
+            "The minimum accuracy level of input A, can be: "
+            "0(unset), 1(float), 2(float16), 3(Bfloat16), 4(int8) (default unset). It is used to control how input A "
+            "is quantized or downcast internally while doing computation, for example: 0 means input A will not be "
+            "quantized or downcast while doing computation, 4 means input A can be quantized with the same block_size"
+            "to int8 internally from type T1.",
+            AttributeProto::INT,
+            static_cast<int64_t>(0))
+        .Attr(
+            "bits",
+            "Number of bits used for weight quantization (default 4)",
+            AttributeProto::INT,
+            static_cast<int64_t>(4))
+        .Attr(
+            "block_size",
+            "Number of groupsize used for weight quantization (default 128)."
+            "It needs to be a power of 2 and not smaller than 16",
+            AttributeProto::INT,
+            static_cast<int64_t>(128))
+        .Input(
+            0,
+            "A",
+            "A 2 dimensional input tensor, not quantized, with shape [M][K]",
+            "T1",
+            OpSchema::Single,
+            true,
+            1,
+            OpSchema::NonDifferentiable)
+        .Input(
+            1,
+            "B",
+            "A 3 dimensional input tensor of shape [N][n_blocks_per_col][blob_size], "
+            "i.e. an N by 2 dimensional data blob containing the packed B bits",
+            "T2",
+            OpSchema::Single,
+            true,
+            1,
+            OpSchema::NonDifferentiable)
+        .Input(
+            2,
+            "scales",
+            "Tensor of quantization scales. It should have shape [N * n_blocks_per_col]",
+            "T1",
```

Comment:
I would not tie scales' type to the activation and output type (i.e. T1). To reduce model size, it is possible to use float16 for scales even when input/output type is float32. There is an example at https://github.com/ggerganov/llama.cpp/blob/f112d198cd9953b633ac662c2705d6e423438717/ggml/src/ggml-common.h#L213 where float16 is used and output is float32.